flake.nix

```diff
@@ -0,0 +1,188 @@
+                inherit system;
+                config.allowBroken = true;
+              };
+
+              individualPackages = builtins.listToAttrs (builtins.map (p: {
+                name = ghcVer + "_" + p;
+                value = builtins.getAttr p newPkgs.haskellPackages;
+              }) categorifierCPackageNames);
+
+              allEnv = let
+                hsenv = newPkgs.haskellPackages.ghcWithPackages (p:
+                  let
+                    deps = builtins.map ({ name, ... }: p.${name})
+                      categorifierCPackages;
+                  in deps);
+              in newPkgs.buildEnv {
+                name = "all-packages";
+                paths = [ hsenv ];
+              };
+            in individualPackages // { "${ghcVer}_all" = allEnv; };
+
+        in packagesOnGHC "ghc8107" // packagesOnGHC "ghc884"
+        // packagesOnGHC "ghc901" // packagesOnGHC "ghc921"
+        // packagesOnGHC "ghcHEAD";
+
+        overlays = fullOverlays;
+
+        devShells = let
+          mkPkgs = ghcVer:
+            let
+              overlayGHC = final: prev: {
+                haskellPackages = prev.haskell.packages.${ghcVer};
+              };
+            in import nixpkgs {
+              overlays = [ overlayGHC (concat.overlay.${system}) ]
+                ++ (categorifier.overlays.${system}) ++ fullOverlays;
+              inherit system;
+              config.allowBroken = true;
+            };
+
+          mkDevShell = ghcVer:
+            let pkgs = mkPkgs ghcVer;
+            in pkgs.haskellPackages.shellFor {
+              packages = ps:
+                builtins.map (name: ps.${name}) categorifierCPackageNames;
+              buildInputs =
+                # use nixpkgs default tools
+                [
+                  pkgs.haskell.packages.ghc8107.cabal-install
+                  pkgs.haskell.packages.ghc8107.hlint
+                ] ++
+                # haskell-language-server on GHC 9.2.1 is broken yet.
+                pkgs.lib.optional (ghcVer != "ghc921")
+                [ pkgs.haskell-language-server ];
+              withHoogle = false;
+            };
+
+          mkUserShell = ghcVer:
+            let
+              pkgs = mkPkgs ghcVer;
+              hsenv = pkgs.haskellPackages.ghcWithPackages
+                (ps: builtins.map (name: ps.${name}) categorifierCPackageNames);
+            in pkgs.mkShell {
+              buildInputs = [ hsenv pkgs.haskellPackages.cabal-install ] ++
+                # haskell-language-server on GHC 9.2.1 is broken yet.
+                pkgs.lib.optional (ghcVer != "ghc921")
+                [ pkgs.haskell-language-server ];
+            };
+
+        in {
+          # nix develop .#ghc8107
+          # (or .#ghc901 .#ghc921)
+          # This is used for building categorifier-c
+          "default" = mkDevShell "ghc901";
+          "ghc8107" = mkDevShell "ghc8107";
+          "ghc901" = mkDevShell "ghc901";
+          "ghc921" = mkDevShell "ghc921";
```

Comment:
Did you try building with 9.2.1? Last time I tried some dependencies didn't yet support 9.2.1

Reply:
now i made a fix to set linear = linear_1_21_7 in this repo. Please check this and let me know if you get GHC 9.2.1 dev shell. :-)

Reply:
> ah right. `linear` dep in categorifier was removed.
> 
> while it seems to be necessary in categorifier-c somehow

My _guess_ is that it should be removed from categorifier-c as well. We're likely providing instances to be used by KH, but those should live downstream.

Either way, not part of this change.

Reply:
I think it was some other library, not linear. But apparently 9.2.1 is tested in CI and it passed, so we have no problem then.